kratos/geometries/nurbs_volume_geometry.h

```diff
@@ -0,0 +1,997 @@
+        SizeType NumPointsPerSpanW) const
+    {
+        auto knot_span_intervals_u = KnotSpanIntervalsU();
+        auto knot_span_intervals_v = KnotSpanIntervalsV();
+        auto knot_span_intervals_w = KnotSpanIntervalsW();
+
+        const SizeType number_of_integration_points =
+            knot_span_intervals_u.size() * knot_span_intervals_v.size() * knot_span_intervals_w.size()
+            * NumPointsPerSpanU * NumPointsPerSpanV * NumPointsPerSpanW;
+
+        if (rIntegrationPoints.size() != number_of_integration_points) {
+            rIntegrationPoints.resize(number_of_integration_points);
+        }
+
+        typename IntegrationPointsArrayType::iterator integration_point_iterator = rIntegrationPoints.begin();
+
+        for (IndexType i = 0; i < knot_span_intervals_u.size(); ++i) {
+            for (IndexType j = 0; j < knot_span_intervals_v.size(); ++j) {
+                for (IndexType k = 0; k < knot_span_intervals_w.size(); ++k) {
+                    IntegrationPointUtilities::IntegrationPoints3D(
+                        integration_point_iterator,
+                        NumPointsPerSpanU, NumPointsPerSpanV, NumPointsPerSpanW,
+                        knot_span_intervals_u[i].GetT0(), knot_span_intervals_u[i].GetT1(),
+                        knot_span_intervals_v[j].GetT0(), knot_span_intervals_v[j].GetT1(),
+                        knot_span_intervals_w[k].GetT0(), knot_span_intervals_w[k].GetT1());
+                }
+            }
+        }
+    }
+
+    Matrix& Jacobian( Matrix& rResult, const CoordinatesArrayType& rCoordinates ) const override
+    {
+        const SizeType working_space_dimension = this->WorkingSpaceDimension();
+        const SizeType local_space_dimension = this->LocalSpaceDimension();
+        const SizeType points_number = this->PointsNumber();
+        if(rResult.size1() != working_space_dimension || rResult.size2() != local_space_dimension)
+            rResult.resize( working_space_dimension, local_space_dimension, false );
+
+        Matrix shape_functions_gradients(points_number, local_space_dimension);
+        ShapeFunctionsLocalGradients( shape_functions_gradients, rCoordinates );
+
+        NurbsVolumeShapeFunction shape_function_container(mPolynomialDegreeU, mPolynomialDegreeV, mPolynomialDegreeW, 1);
+        shape_function_container.ComputeBSplineShapeFunctionValues(mKnotsU,mKnotsV,mKnotsW,
+            rCoordinates[0], rCoordinates[1], rCoordinates[2]);
+        std::vector<array_1d<int,3>> indices = shape_function_container.NonzeroControlPointIndices();
+        SizeType number_of_cp = shape_function_container.NumberOfNonzeroControlPoints();
+
+        SizeType number_cp_u = NurbsUtilities::GetNumberOfControlPoints(mPolynomialDegreeU, mKnotsU.size());
+        SizeType number_cp_v = NurbsUtilities::GetNumberOfControlPoints(mPolynomialDegreeV, mKnotsV.size());
+        SizeType number_cp_w = NurbsUtilities::GetNumberOfControlPoints(mPolynomialDegreeW, mKnotsW.size());
+
```

Comment:
are you sure? isn't it also used to compute the shape functions?

Reply:
No that happens a bit earlier inside the `ShapeFunctionsLocalGradients` function. Inside that function we compute the shape functions.

Reply:
So as it is right now we instantiate the `shape_function_container` twice. Therefore I would suggest make the `NonzeroControlPointIndices` static and put all required operations in there?

Reply:
Well one more thought to this. Actually right now, we have two functions in the shape functions class: One is called `NonZeroControlPointIndices` and the other is `ControlPointIndices`. However, to my understanding they do exactly the same, except that the former version returns the indices in all three space directions and the `ControlPointIndices` returns the actual index (However still only for the NonzeroCP's). So what do you think if we get rid of one implementation and provide only one that is called `NonzeroControlPointIndices` and which will return the actual indices in a `vector<int> `.
We could make it static and change the interface to NonzeroControlPointIndices(rPolyDegreeU-W, rKnotsU-W, rParameterU-W, NumberOfControlPointsU-W), because all that information is required to compute the indices. Another option would be to add this function to the `nurbs_utilities.h`?